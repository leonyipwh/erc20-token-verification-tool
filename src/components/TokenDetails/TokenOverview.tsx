import { useEffect, useState, useContext } from 'react';
import TokenContext from '../../context/tokenFormContext';
import Loader from '../Loader';
import { getAbi } from '../../api/etherscan';
import { fetchProfile } from '../../helpers/token/tokenProfileHelper';
import Profile from '../TokenProfile';
import GasEstimation from '../GasEstimation';
import Grid from '@mui/material/Grid';

interface ITokenProfile {
  label: string;
  value: number | string | undefined;
}

const TokenOverview = () => {
  const [hasError, setHasError] = useState<string | null>();
  const [tokenProfile, setTokenProfile] = useState<ITokenProfile[]>([]);
  const { submitForm } = useContext(TokenContext);

  useEffect(() => {
    setHasError(null);
    setTokenProfile([]);
    getTokenProfile();
  }, [submitForm]);

  const getTokenProfile = async () => {
    const { network, address, provider } = submitForm;
    if (!provider) return;
    try {
      const { result } = await getAbi(network, address);
      const profileData = await fetchProfile(address, result, provider);

      setTokenProfile(profileData);
    } catch (err) {
      console.error(err);
      const { result } = err;
      setHasError(result);
    }
  };

  return (
    <>
      {hasError && <>{hasError}</>}
      {!hasError && !tokenProfile.length ? (
        <Loader />
      ) : (
        !hasError && (
          <Grid
            container
            direction="column"
            spacing={3}
          >
            <Grid item>
              <Profile data={tokenProfile} />
            </Grid>
            <Grid item>
              <GasEstimation />
            </Grid>
          </Grid>
        )
      )}
    </>
  );
};

export default TokenOverview;
