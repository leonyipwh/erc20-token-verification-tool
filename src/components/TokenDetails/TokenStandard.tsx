import { useEffect, useState, useContext } from 'react';
import { getAbi } from '../../api/etherscan';
import Loader from '../Loader';
import { compatibilityCheck } from '../../helpers/abi/abiHelper';
import CompatibilityResult from '../CompatibilityResult';
import TokenContext from '../../context/tokenFormContext';
import SnackbarContext from '../../context/snackbarContext';
import { multiPropDiff } from '@netilon/differify';

const TokenStandard = () => {
  const [loading, setLoading] = useState<boolean>(true);
  const [abiScanOutput, setAbiScanOutput] = useState<multiPropDiff[]>([]);
  const { submitForm } = useContext(TokenContext);
  const { snackbarState, setSnackbarState } = useContext(SnackbarContext);

  useEffect(() => {
    setLoading(true);
    fetchTokenDetails();
  }, [submitForm]);

  const fetchTokenDetails = async () => {
    const { network, address } = submitForm;
    try {
      const { status, message, result } = await getAbi(network, address);

      if (status !== '1') throw new Error(message);

      setAbiScanOutput(compatibilityCheck(result));
    } catch (err) {
      console.error(err);
      setSnackbarState({
        ...snackbarState,
        open: true,
        text: 'Unable to get ABI'
      });
    } finally {
      setLoading(false);
    }
  };

  return <>{loading ? <Loader /> : <CompatibilityResult data={abiScanOutput} />}</>;
};

export default TokenStandard;
