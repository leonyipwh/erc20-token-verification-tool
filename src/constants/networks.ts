import { INetwork } from '../types/tokenVerification';

export const networks: INetwork[] = [
  {
    name: 'mainnet',
    subdomain: 'api'
  },
  {
    name: 'rinkeby',
    subdomain: 'api-rinkeby'
  },
  {
    name: 'ropsten',
    subdomain: 'api-ropsten'
  },
  {
    name: 'goerli',
    subdomain: 'api-goerli'
  }
];
