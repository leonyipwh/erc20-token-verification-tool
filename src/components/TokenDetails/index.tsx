import React, { useContext } from 'react';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import Box from '@mui/material/Box';
import TokenContext from '../../context/tokenFormContext';
import TokenOverview from '../../components/TokenDetails/TokenOverview';
import TokenStandard from '../../components/TokenDetails/TokenStandard';

interface TabPanelProps {
  children?: React.ReactNode;
  index: number;
  value: number;
}

function TabPanel(props: TabPanelProps) {
  const { children, value, index, ...other } = props;

  return (
    <div
      role="tabpanel"
      hidden={value !== index}
      id={`simple-tabpanel-${index}`}
      aria-labelledby={`simple-tab-${index}`}
      {...other}
    >
      {value === index && <Box sx={{ p: 3 }}>{children}</Box>}
    </div>
  );
}

function a11yProps(index: number) {
  return {
    id: `simple-tab-${index}`,
    'aria-controls': `simple-tabpanel-${index}`
  };
}

const TokenDetails = () => {
  const [value, setValue] = React.useState(0);
  const { submitForm } = useContext(TokenContext);

  const handleChange = (event: React.SyntheticEvent, newValue: number) => {
    setValue(newValue);
  };

  const tabList = [
    {
      label: 'Overview',
      component: <TokenOverview />
    },
    {
      label: 'Token Standard',
      component: <TokenStandard />
    }
  ];

  return submitForm?.address ? (
    <Box sx={{ width: '100%', mt: 1 }}>
      <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
        <Tabs
          value={value}
          onChange={handleChange}
          aria-label="token detail tabs"
        >
          {tabList.map((item, index) => (
            <Tab
              key={index}
              label={item.label}
              {...a11yProps(index)}
            />
          ))}
        </Tabs>
      </Box>
      {tabList.map((item, index) => (
        <TabPanel
          value={value}
          index={index}
          key={index}
        >
          {item.component}
        </TabPanel>
      ))}
    </Box>
  ) : (
    <></>
  );
};

export default TokenDetails;
