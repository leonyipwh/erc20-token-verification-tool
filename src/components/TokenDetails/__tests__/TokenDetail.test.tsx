import TokenDetails from '../index';
import { render } from '@testing-library/react';
import TokenFormContext from '../../../context/tokenFormContext';
import { networks } from '../../../constants/networks';

jest.mock('../../../components/TokenDetails/TokenOverview', () => {
  const TokenOverview = () => <>Token Overview</>;
  return TokenOverview;
});

jest.mock('../../../components/TokenDetails/TokenStandard', () => {
  const TokenStandard = () => <>Token Standard</>;
  return TokenStandard;
});

describe('<TokenDetails>', () => {
  it('should render', () => {
    const submitForm = {
      network: networks[0],
      address: 'add'
    };
    const { asFragment } = render(
      <TokenFormContext.Provider value={{ submitForm, setSubmitForm: () => {} }}>
        <TokenDetails />
      </TokenFormContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render empty fragment', () => {
    const { asFragment } = render(<TokenDetails />);
    expect(asFragment()).toMatchSnapshot();
  });
});
