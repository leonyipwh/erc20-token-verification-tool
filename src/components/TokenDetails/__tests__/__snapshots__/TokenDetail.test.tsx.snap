// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenDetails> should render 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-8uw9dg"
  >
    <div
      class="MuiBox-root css-1gsv261"
    >
      <div
        class="MuiTabs-root css-1ujnqem-MuiTabs-root"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            aria-label="token detail tabs"
            class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
            role="tablist"
          >
            <button
              aria-controls="simple-tabpanel-0"
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
              id="simple-tab-0"
              role="tab"
              tabindex="0"
              type="button"
            >
              Overview
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-controls="simple-tabpanel-1"
              aria-selected="false"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
              id="simple-tab-1"
              role="tab"
              tabindex="-1"
              type="button"
            >
              Token Standard
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <span
            class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
    </div>
    <div
      aria-labelledby="simple-tab-0"
      id="simple-tabpanel-0"
      role="tabpanel"
    >
      <div
        class="MuiBox-root css-19kzrtu"
      >
        Token Overview
      </div>
    </div>
    <div
      aria-labelledby="simple-tab-1"
      hidden=""
      id="simple-tabpanel-1"
      role="tabpanel"
    />
  </div>
</DocumentFragment>
`;

exports[`<TokenDetails> should render empty fragment 1`] = `<DocumentFragment />`;
